### PR TITLE
 YOLO Family: Detection Generator Base Operation PR (1/2)

### PR DESCRIPTION
**Pull request by vishnubanna**
# Description
This PR updates all the ops needed for the YOLO detection generator and adds an nms_ops file that holds the nms used for box filtering on the TPU. This PR adds DIOU nms as well as a quick NMS that is used for compiling the model in tflite.   

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Documentation update
- [ ] TensorFlow 2 migration
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] A new research paper code implementation
- [ ] Other (Specify)

## Tests

box op and math op test cases 
```
python3.8 -m official.vision.beta.projects.yolo.ops.box_ops_test
```

nms op tests will be included in the next commit

**Test Configuration**:

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by vishnubanna** [comment]
@googlebot I consent.

**Comment by anivegesana** [comment]
@googlebot I consent.

**Comment by The-Indian-Chinna** [comment]
@googlebot I consent.

**Comment by vishnubanna** [comment]
@googlebot I consent.

File: official/vision/beta/projects/yolo/utils/math_ops.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+"""A set of private math operations used to safely implement the YOLO loss."""
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/utils/math_ops.py:1]: Can you revert? I thought you'll rename math_ops to utils. But I feel math_ops name is fine.

Comment (vishnubanna) [reply official/vision/beta/projects/yolo/utils/math_ops.py:1]: yes we can revert this.

File: official/vision/beta/projects/yolo/ops/nms_ops.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
+import tensorflow as tf
+from official.vision.beta.ops import box_ops as box_utils
+from official.vision.beta.projects.yolo.ops import box_ops as box_ops
+
+NMS_TILE_SIZE = 512
+
+class TiledNMS():
```
[the remark this answers is not in the document]
Comment (arashwan) [reply official/vision/beta/projects/yolo/ops/nms_ops.py:7]: Also why do you have this as a class? Can you just use free functions to be consistent with model garden implementation?

File: official/vision/beta/projects/yolo/ops/nms_ops.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+
+  def __init__(self, iou_type='diou', beta=0.6):
+    '''Initialization for all non max suppression operations mainly used to
+    select hyperparameters for the iou type and scaling.
+
+    Args:
+      iou_type: `str` for the version of IOU to use {diou, giou, ciou, iou}.
+      beta: `float` for the amount to scale regularization on distance iou.
+    '''
+    self._iou_type = TiledNMS.IOU_TYPES[iou_type]
```
Comment (arashwan) [inline official/vision/beta/projects/yolo/ops/nms_ops.py:18]: Maybe use this an argument, and make the nms a free function not a class?

File: official/vision/beta/projects/yolo/README.md [hunk 4]
```diff
@@ -17,30 +17,31 @@ repository.
 
 ## Description
 
-Yolo v1 the original implementation was released in 2015 providing a ground
-breaking algorithm that would quickly process images, and locate objects in a
-single pass through the detector. The original implementation based used a
-backbone derived from state of the art object classifier of the time, like
+YOLO v1 the original implementation was released in 2015 providing a 
+groundbreakingalgorithm that would quickly process images and locate objects in
```
[the remark this answers is not in the document]
Comment (The-Indian-Chinna) [reply official/vision/beta/projects/yolo/README.md]: Addressed in https://github.com/tensorflow/models/pull/10040/commits/836d599fa181d0c89c3ffc8c2177f567a3bbedfb.

File: official/vision/beta/projects/yolo/ops/nms_ops.py [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+from official.vision.beta.ops import box_ops as box_utils
+from official.vision.beta.projects.yolo.ops import box_ops as box_ops
+
+NMS_TILE_SIZE = 512
+
+class TiledNMS():
+  IOU_TYPES = {'diou': 0, 'giou': 1, 'ciou': 2, 'iou': 3}
+
+  def __init__(self, iou_type='diou', beta=0.6):
+    '''Initialization for all non max suppression operations mainly used to
```
[the remark this answers is not in the document]
Comment (The-Indian-Chinna) [reply official/vision/beta/projects/yolo/ops/nms_ops.py]: Addressed in https://github.com/tensorflow/models/pull/10040/commits/836d599fa181d0c89c3ffc8c2177f567a3bbedfb.

File: official/vision/beta/projects/yolo/ops/nms_ops.py [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+def sorted_non_max_suppression_padded(scores, boxes, max_output_size,
+                                      iou_threshold):
+  """wrapper function to match NMS found in official/vision/beta/ops/nms.py"""
+  return BASE_NMS._sorted_non_max_suppression_padded(scores, boxes,
+                                                     max_output_size,
+                                                     iou_threshold)
+
+
+def sort_drop(objectness, box, classificationsi, k):
+  """This function sorts and drops boxes such that there are only k boxes
```
[the remark this answers is not in the document]
Comment (The-Indian-Chinna) [reply official/vision/beta/projects/yolo/ops/nms_ops.py]: Addressed in https://github.com/tensorflow/models/pull/10040/commits/836d599fa181d0c89c3ffc8c2177f567a3bbedfb.

**Review by arashwan — DISMISSED** [review]

**Review by jaeyounkim — APPROVED** [review]

**Review by jaeyounkim — APPROVED** [review]

**Comment by jaeyounkim** [comment]
This PR is already merged.

**Pull request closed without merging**